### PR TITLE
HGCal - Remove ghost detids

### PR DESCRIPTION
**Pull request by felicepantaleo**
This PR is a fix for a hopefully temporary bug in the geometry.
In the geometry we have some hidden, ghost detIds that should not exist. 
You can find the list here: www.cern.ch/felice.pantaleo/ghosts.txt

### CODE REVIEW

**Comment by felicepantaleo** [comment]
type bug fix

**Comment by felicepantaleo** [comment]
FYI: @malgeri @rovere @cseez @kpedro88 @vandreev11

File: RecoLocalCalo/HGCalRecProducers/python/HGCalRecHit_cfi.py [hunk 1]
```diff
@@ -75,6 +75,13 @@
     HGCHEF_fCPerMIP = HGCalUncalibRecHit.HGCHEFConfig.fCPerMIP,
     HGCHEB_keV2DIGI = hgchebackDigitizer.digiCfg.keV2MIP,
     HGCHEB_isSiFE   = HGCalUncalibRecHit.HGCHEBConfig.isSiFE,
+    # don't produce rechit if detid is a ghost one
+    # smallest and largest ghost detid on positive side
+    GhostDetIdPosMin = 1162362881,
+    GhostDetIdPosMax = 1162362952,
```
Comment (slava77) [inline RecoLocalCalo/HGCalRecProducers/python/HGCalRecHit_cfi.py]: @bsunanda 
are you aware of these?

**Comment by slava77** [comment]
@felicepantaleo 
https://github.com/cms-sw/cmssw/pull/www.cern.ch/felice.pantaleo/ghosts.txt does not point to anything . maybe there is a typo in the link.

What does the geometry return for these detids?
If they are in non-existent locations, perhaps it's better to check that.

If the origin of these detids is not understood, how can one predict that the set made in this PR would apply to a newly created samples or some other existing samples that you perhaps did not look at yet.

**Comment by felicepantaleo** [comment]
@slava77 , I updated the link, now it should be correct.

Basically there is some "remnant" of a layer inside the absorber (!). It is a bug in the geometry, which is still not fixed (and I would not know how to fix it).
You can get hits, coordinates and everything, just like it was a real sensitive layer.
 So this is a temporary workaround since the geometry won't be fixed before the MC production...
@bsunanda and @malgeri can comment

**Comment by slava77** [comment]
@felicepantaleo 
can these bad detids be captured with a detid mask or even from an index known to HGC det id (like layer)?
This would be a more straightforward choice than magic numbers.

**Comment by felicepantaleo** [comment]
@slava77 
I cannot exclude that, if there is a mask containing all of them, this mask does not include also valid detIds.

**Comment by felicepantaleo** [comment]
what I see is that they all have the same z, and at this z-coordinate there should be no layer...
But IMHO this is a weak condition...

**Comment by slava77** [comment]
@bsunanda 
please clarify if the bug in the HGC geometry is understood already and how long it may take to fix it.

File: RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitWorkerSimple.h [hunk 2]
```diff
@@ -47,6 +47,11 @@ class HGCalRecHitWorkerSimple : public HGCalRecHitWorkerBaseClass {
   std::vector<int> v_DB_reco_flags_;
   bool killDeadChannels_;
 
+  uint32_t GhostDetIdPosMin_;
+  uint32_t GhostDetIdPosMax_;
+  uint32_t GhostDetIdNegMin_;
+  uint32_t GhostDetIdNegMax_;
```
Comment (slava77) [inline RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitWorkerSimple.h]: start local variables and data members with lower case letter. Capitalization is reserved for class names.

File: RecoLocalCalo/HGCalRecProducers/python/HGCalRecHit_cfi.py (hunk 1 above)
Comment (slava77) [inline RecoLocalCalo/HGCalRecProducers/python/HGCalRecHit_cfi.py]: capitalization rule applies to python configs as well, although it is not enforced particularly strictly

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitWorkerSimple.cc [hunk 3]
```diff
@@ -87,6 +91,10 @@ bool HGCalRecHitWorkerSimple::run(const edm::Event & evt, const HGCUncalibratedR
         HGCRecHitCollection & result)
 {
     DetId detid = uncalibRH.id();
+// don't produce rechit if detid is a ghost one
+    if((detid < GhostDetIdPosMax_ +1  and GhostDetIdPosMin_-1 < detid ) or (detid < GhostDetIdNegMax_+1  and GhostDetIdNegMin_-1 < detid ) )
```
Comment (slava77) [inline RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitWorkerSimple.cc]: there is no check that the IDs are in the positive or negative eta.
As such the semantics of "Pos" and "Neg" is not significant and if the same feature persists the code can be extended to look in ranges of det Ids.

File: RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitWorkerSimple.cc [hunk 4]
```diff
@@ -54,7 +54,11 @@ HGCalRecHitWorkerSimple::HGCalRecHitWorkerSimple(const edm::ParameterSet&ps) :
     hgcHEF_noise_fC_ = ps.getParameter < std::vector<double> > ("HGCHEF_noise_fC");
     hgcHEB_noise_MIP_ = ps.getParameter<double>("HGCHEB_noise_MIP");
 
-
+    // don't produce rechit if detid is a ghost one
+    GhostDetIdPosMin_ = ps.getParameter<int>("GhostDetIdPosMin");
+    GhostDetIdPosMax_ = ps.getParameter<int>("GhostDetIdPosMax");
+    GhostDetIdNegMin_ = ps.getParameter<int>("GhostDetIdNegMin");
```
Comment (slava77) [inline RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitWorkerSimple.cc]: the underlying type of DetId is uint32. Better keep it this way here.

**Comment by slava77** [comment]
Looking at the min/max IDs,
a bit more readable in hex they are
- pos eta ```0x45484001 - 0x45484048```
- neg eta ```0x45404001 - 0x45404048 ```

These translate to HCAL, endcap, in new format, depth 4, [z side 1/0], eta 16, phi 1 to 72 [0x48].
Assuming that you do not need any other phi values at this eta and depth, the one line check becomes ``` detid & 0xFFF7FE00 == 0x45404000 ``` [or ```detid & 4294442496U == 1161838592```]

File: RecoLocalCalo/HGCalRecProducers/python/HGCalRecHit_cfi.py [hunk 5]
```diff
@@ -75,6 +75,13 @@
     HGCHEF_fCPerMIP = HGCalUncalibRecHit.HGCHEFConfig.fCPerMIP,
     HGCHEB_keV2DIGI = hgchebackDigitizer.digiCfg.keV2MIP,
     HGCHEB_isSiFE   = HGCalUncalibRecHit.HGCHEBConfig.isSiFE,
+    # don't produce rechit if detid is a ghost one
+    # smallest and largest ghost detid on positive side
+    GhostDetIdPosMin = 1162362881,
```
Comment (slava77) [inline RecoLocalCalo/HGCalRecProducers/python/HGCalRecHit_cfi.py]: run time errors are not surprisingly coming from here: there is no cms type assigned to the values

File: RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitWorkerSimple.cc (hunk 3 above)
Comment (slava77) [inline RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitWorkerSimple.cc]: as mentioned in the main PR thread, it seems simpler and perhaps more appropriate to use just two values , let's call them rangeMask and rangeMatch and use ```detid & rangeMask == rangeMatch ``` as a check.
It's pure luck that phi is  in the least significant bits to rely on simple int range matching.

**Comment by felicepantaleo** [comment]
thanks @slava77 for always giving precious advices

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/20008	bba80ec
- changes in the code are in line with the description. This is OK for a temporary solution; although a more straightforward approach eventually would be to remove this at the origin or at least in the digitizer. Given that there are some existing samples on which reco can be run for studies, the fix in this PR is fine.
- jenkins tests pass and comparisons with baseline show differences in the 2023 workflows starting from HGC HEB rechits at the innermost eta, as expected.

looking at wf 20034 in cms-bot outputs:

there is some reduction in the number of hits in HGCHEB, as expected from the detid values 
all_oldvsnew_ttbar14tev2023d17wf20034p0c_log10hgcrechitssorted_hgcalrechit_hgchebrechits_reco_obj_obj_energy

this eventually percolates down to higher level reco
curiously, at the supercluster level, there are more SCs
all_oldvsnew_ttbar14tev2023d17wf20034p0c_recocaloclusters_particleflowsuperclusterhgcal__reco_obj_eta
The changes are in the right place (eta 16)

I notice that we have "null" PF clusters (we don't have them in layerClusters or superclusters)
all_oldvsnew_ttbar14tev2023d17wf20034p0c_recopfclusters_particleflowclusterhgcal__reco_obj_eta
This could be something to cleanup or assign properly in some follow up update.

**Comment by slava77** [comment]
@kpedro88 
a signature from upgrades is missing.
Please check this PR and suggest changes or sign it.
Thank you.

**Comment by kpedro88** [comment]
+1

**Comment by davidlange6** [comment]
hi- just to confirm - is @bsunanda or someone else working or planning to work on a fix later?

**Comment by felicepantaleo** [comment]
hi @davidlange6 , yes the plan is to review the geometry after the TDR, and this problem will be solved.

**Comment by davidlange6** [comment]
+1

**Pull request merged**